### ISSUE TITLE
new feature? -->event registration deadline

### ISSUE THREAD

**Issue by jojo12** [jemproject/JEM-Project#1629]
We have a deadline for unregistration, but not for event registration.
Of course it is possible to create an event field for this or it  is possible to write the deadline into the description.
Somehow it would be good to add this as a new feature. What do you think? 
EDIT:
or can this be the same? deadline to register and unregister?

**Comment by mckillo** [comment]
I would like the idea. You can create an event and open registration for a specific date. One proposal to v.4.1.0

**Comment by Wolfgang-Chess** [comment]
I agree. Good proposal.

**Comment by mckillo** [comment]
Proposal change to add open registratation before starts event

![image](https://github.com/user-attachments/assets/528f3133-425d-45da-a511-7a9b1ac47499)

**Comment by mckillo** [comment]
These fields (From, Until) are currently in hours, but I think it would be better to have them in minutes for more flexible time settings.
What do you opinion?

**Comment by jojo12** [comment]
I agree for 4.3.1 
Always when it's near a frontier, you're very creative😉

**Comment by mckillo** [comment]
Example: This is how the event is created with the registration settings (active from x hours before):

![image](https://github.com/user-attachments/assets/3be13241-fa95-472b-ab46-b1e64ed206bf)

The new field (from) is a number with decimal ( float from 0 to 9999,9999 hours).

Frontend view: At the bottom of the event page, you can see:

![image](https://github.com/user-attachments/assets/0617fdbc-7556-46ab-818e-c4151c769c65)

When registration is available, the user will see this:

![image](https://github.com/user-attachments/assets/c466529f-c878-4031-808e-cb4c97f8d889)

If the user is logged in, will see:

![image](https://github.com/user-attachments/assets/95348c1f-cc20-4f04-a031-30f68cf0fc2f)

This new feature will upload a new branch 'event_registration_hours'

**Comment by mckillo** [comment]
I have changed the field to this format HHHH:MM (up to 9999:99 for hours and minutes).
I believe this makes it easier for the user to enter the value.

![image](https://github.com/user-attachments/assets/66b59464-ddf4-40c0-84af-35d12e001d58)

**Comment by mckillo** [comment]
Example for this new feature to understarnd the function in the time (in this demo, all processes must be completed within  1 minute, it's minimum time in the setting, before JEM 4.3 only has hours in annulation). 

1. Set a new event with registration

![image](https://github.com/user-attachments/assets/7d83186b-2903-4d3f-81e2-ececec290dab)

The registration starts 5 minutes before to event begins. The event starts to 17:00, so the registration form will be available until 16:55. If the user is registered, they can cancel their  registration up to 4 minutes before the event starts, which mean until 16:56

**Comment by mckillo** [comment]
Screenshots of what the user sees in the event view (FE)

**At 16:54:**
![image](https://github.com/user-attachments/assets/9d580c1f-fe4c-4ae8-a2f4-76da4d1c1f71)

**At 16:55:**
![image](https://github.com/user-attachments/assets/51f93e30-9404-49e4-9fe8-c78245b5af92)
And after registration
![image](https://github.com/user-attachments/assets/771d67bd-52fd-46ad-b03e-45747a974305)

**At 16:56:**
![image](https://github.com/user-attachments/assets/0036f344-643e-4fa1-a0ab-addda09fc88c)
In this case, the options are disabled.

Ready to test it.

**Comment by jojo12** [comment]
I tested: looks good. 
There is a start point for event registration,  there is an end point for event cancelling
but is there a end point for event registration? or is the end point for event cancelling the same?

**Comment by mckillo** [comment]
These fields are in hours to maintain compatibility with older events that have cancellation settings for a specified number of hours before the event starts.

**Comment by mckillo** [comment]
We currently don't have an endpoint for event registration, and I expect there will be many dates. I prefer to have a stable process for the new this initial feature before expanding it.

I think that it's not necessary to expand the setting. If there is available places for an event, why limit it?. It's better for users to register (and the organization) allow registration even if the event has already started (normal). If there are no places, then registration process shouldn't allow them to register or it should pull from the waitlist.

**Comment by mckillo** [comment]
is there a possibility to localize the date/time?
.... other alternative, I can change the input (float of hours) to input type="datetime-local" ... more visual.
Lets me made it and do the data convert in hours (during the weekend)

**Comment by jojo12** [comment]
We have the following situation. For a weekend in an external place, we have to announce to the hostel, how many beds we need. We have to communicate this two weeks before the start of the weekend. Better we reserved 20 beds. We have to communicate two weeks before the weekend, if we need all beds or we can give free some. If we don't communicate two weeks before, we have to pay all beds, even if we need only ten.
In this case we need an end point of registration!

**Comment by mckillo** [comment]
Come on only one more :-)
Add it.

**Comment by mckillo** [comment]
![image](https://github.com/user-attachments/assets/4d9437e2-0f8d-4b86-a400-b8ac206d6a10)

**Comment by jojo12** [comment]
Excellent proposal! But you don't have time until weekend! (And you have a first life!)

**Comment by mckillo** [comment]
Yes...copy & paste code (xml and ini file)  to see if it's right.
This weekend, I'll finish and test the last RC.

**Comment by mckillo** [comment]
Here you are the new field 'registra_until' is included to specify the end time of registration for the event. 
This field follows the same format: HHHH:MM or HHHH for hours and minutes (for the moment).

![image](https://github.com/user-attachments/assets/cacfd145-c5f0-412f-a1c8-3ed44c4b22f2)

When the registration is open:

![image](https://github.com/user-attachments/assets/3487cee8-9fcc-4ccd-9e66-e07828c56f87)

The user can complete the registration form until  05:26

![image](https://github.com/user-attachments/assets/13a89536-37bb-4d5b-939b-ee1f0608def8)

From 05:26 to 05:27, the user can cancel the registration. However, if the user cancels, can't register again, because the registration form (options and bottom) is disabled after 05:26.

![image](https://github.com/user-attachments/assets/35127164-a413-4d9c-8228-ba0af1733ac4)

Play around with it and tell me if it’s good.

**Comment by mckillo** [comment]
And if you have JEM 4.3.0 (any RC) installed, when install this commit, don't forget to apply 4.3.1.sql

**Comment by mckillo** [comment]
Last commit: Settings and details in the view style (frontend) for responsive and legacy layout.

**Comment by mckillo** [comment]
**Questions-proposal**
![image](https://github.com/user-attachments/assets/bac134d6-c754-4a44-a3a1-b124fe904003)

**Comment by mckillo** [comment]
The proposal can be a long text. Other short: 
_Open (from): Format HHHH:MM
Close (until): Format HHHH:MM_

For the moment, I change it to 
![image](https://github.com/user-attachments/assets/6b82dcc6-70a6-4e97-b678-164a8708aba2)

There are options that require some thought.

**Comment by mckillo** [comment]
Change the type of these fields to calendar:

![image](https://github.com/user-attachments/assets/a077775a-f936-4fad-ac87-a2a0e232439f)

**Comment by mckillo** [comment]
Enjoy with second version of this issue. Now, JEM uses datetime in these fields, making it easier for users.
Check it when you have time.

**Comment by jojo12** [comment]
I tried to copy the files inside my localhost
in frontend I get:
Error
The form cannot be submitted as it's missing required data.
Please correct the marked fields and try again.
-->But there is nothing marked.

**Comment by mckillo** [comment]
Send me a capture the event setting

**Comment by jojo12** [comment]
Backend:
![grafik](https://github.com/user-attachments/assets/fb51482a-17d3-4f4c-9591-5b3925dbcc7f)

**Comment by jojo12** [comment]
frontend:
![grafik](https://github.com/user-attachments/assets/fa190741-fbb4-4e0b-9c3b-d05198c5fa78)

**Comment by jojo12** [comment]
![grafik](https://github.com/user-attachments/assets/9cd73e0b-a89d-4a41-9320-2d65bffb9006)

**Comment by mckillo** [comment]
Sorry, I have updated the 4.3.1.sql, because your setting is for hours, not datetime. I upload a new commit about it.
Thanks to report it.

**Comment by mckillo** [comment]
You can delete in the structure event table: regsitra_until, registra_from, unregistra_until and run the lines:
```

ALTER TABLE `#__jem_events` ADD COLUMN `registra_until` datetime DEFAULT NULL AFTER `registra`;
ALTER TABLE `#__jem_events` ADD COLUMN `registra_from` datetime DEFAULT NULL AFTER `registra`;
ALTER TABLE `#__jem_events` CHANGE `unregistra_until` `unregistra_until` datetime DEFAULT NULL;
```

**Comment by mckillo** [comment]
Please check the commit and update the events.xml in both the front end and back end. The new fields are not required. There are events that dont't use from and until in the registrations.

**Comment by jojo12** [comment]
Welldone!
Will test more later, I have to do a transfer.

**Comment by mckillo** [comment]
Ok, thanks. 

Once 4.3.0 is released and @Heklaterriol creates a new JEM-4.3.1-dev branch, the event_registration_hours branch can be merged into JEM-4.3.1-dev, delete event_registration_hours branch and this issue can be closed.
